### PR TITLE
Fixup margins on the Android Box binding

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Comment by swankjesse** [comment]
https://github.com/cashapp/redwood/assets/133019/d67574ec-19af-4d9c-9a00-5ffece237d4f

File: redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt [hunk 1]
```diff
@@ -36,66 +34,91 @@ import app.cash.redwood.layout.widget.Box
...
 
-  override val value = object : FrameLayout(context) {
-    init {
-      layoutParams = LayoutParams(MATCH_PARENT, MATCH_PARENT)
+  init {
+    layoutParams = LayoutParams(MATCH_PARENT, MATCH_PARENT)
+  }
 
-      // TODO: Remove this.
-      setBackgroundColor(0xFFFFFF66.toInt())
```
Comment (swankjesse) [inline redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt:57]: FYI, I did this AFTER recording the video!

File: test-app/presenter/src/commonMain/kotlin/com/example/redwood/testing/presenter/BoxSandbox.kt [hunk 2]
```diff
@@ -203,17 +203,23 @@ private fun ColumnScope.BoxRow(
           Rectangle(
             backgroundColor = backColor,
             cornerRadius = 12f,
-            modifier = Modifier.width((rowHeight.value * 1.4).dp).height((rowHeight.value * 0.8).dp),
+            modifier = Modifier.width((rowHeight.value * 1.4).dp)
+              .height((rowHeight.value * 0.8).dp)
+              .margin(Margin(5.dp)),
```
Comment (swankjesse) [inline test-app/presenter/src/commonMain/kotlin/com/example/redwood/testing/presenter/BoxSandbox.kt:208]: I added margins to the sample!

File: redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt [hunk 3]
```diff
@@ -36,66 +34,91 @@ import app.cash.redwood.layout.widget.Box
...
-    }
+  override fun width(width: Constraint) {
+    this.width = width
+  }
 
-    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
```
Comment (swankjesse) [inline redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt:60]: This isn’t deleted; it’s just moved.

File: redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt [hunk 4]
```diff
@@ -36,66 +34,91 @@ import app.cash.redwood.layout.widget.Box
...
+      layoutParams.setFrom(child.modifier)
+      child.value.layoutParams = layoutParams // To force layout.
     }
-  }
 
-  override val children: ViewGroupChildren = ViewGroupChildren(value)
+    // If we're supposed to fill, never measure a size smaller than what's offered.
+    // (This will turn MeasureSpec.AT_MOST into MeasureSpec.EXACTLY.)
+    if (width == Constraint.Fill) widthMode = MeasureSpec.EXACTLY
+    if (height == Constraint.Fill) heightMode = MeasureSpec.EXACTLY
```
Comment (swankjesse) [inline redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt:104]: This part is new, and replaces the previous call to `setMeasuredDimension()`. We don’t really wanna call that here 'cause the call to `super.onMeasure()` will do that also

File: redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt [hunk 5]
```diff
@@ -36,66 +34,91 @@ import app.cash.redwood.layout.widget.Box
...
-import kotlin.math.roundToInt
 
 internal class ViewBox(
   context: Context,
-) : Box<View> {
+) : FrameLayout(context), Box<View> {
+  override var modifier: Modifier = Modifier
 
-  private val density = Density(context.resources)
+  override val value = this
```
Comment (JakeWharton) [inline redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt:43]:
```suggestion
  override val value: Box<View> get() = this
```
Saves a backing field and removes a covariant override

File: redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt [hunk 6]
```diff
@@ -36,66 +34,91 @@ import app.cash.redwood.layout.widget.Box
...
+  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+    var widthMode = MeasureSpec.getMode(widthMeasureSpec)
+    val widthSize = MeasureSpec.getSize(widthMeasureSpec)
+    var heightMode = MeasureSpec.getMode(heightMeasureSpec)
+    val heightSize = MeasureSpec.getSize(heightMeasureSpec)
+
+    for (child in children.widgets) {
+      val layoutParams = child.value.layoutParams as LayoutParams
+      layoutParams.setFrom(child.modifier)
+      child.value.layoutParams = layoutParams // To force layout.
```
Comment (JakeWharton) [inline redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt:98]:
```suggestion
      child.value.updateLayoutParams<LayoutParams> {
        setFrom(child.modifier)
      }
```
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:core/core-ktx/src/main/java/androidx/core/view/View.kt;l=326-332?q=updateLayoutParams&start=11

**Review by JakeWharton — APPROVED** [review]

File: redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt [hunk 7]
```diff
@@ -36,66 +34,91 @@ import app.cash.redwood.layout.widget.Box
...
-import kotlin.math.roundToInt
 
 internal class ViewBox(
   context: Context,
-) : Box<View> {
+) : FrameLayout(context), Box<View> {
+  override var modifier: Modifier = Modifier
 
-  private val density = Density(context.resources)
+  override val value: Box<View> get() = this
```
Comment (JakeWharton) [inline redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/ViewBox.kt]:
```suggestion
  override val value: View get() = this
```
Sorry I got it wrong first time. This is a `Widget<View>`'s `value` which would be typed `View`. 🤦‍♂️

File: redwood-lazylayout-view/src/main/kotlin/app/cash/redwood/lazylayout/view/ViewLazyList.kt [hunk 8]
```diff
@@ -131,10 +131,10 @@ internal open class ViewLazyList private constructor(
...
     with(density) {
       recyclerView.updatePaddingRelative(
-        start = margin.start.toPx().toInt(),
-        top = margin.top.toPx().toInt(),
-        end = margin.end.toPx().toInt(),
-        bottom = margin.bottom.toPx().toInt(),
+        start = margin.start.toPxInt(),
+        top = margin.top.toPxInt(),
+        end = margin.end.toPxInt(),
+        bottom = margin.bottom.toPxInt(),
```
Comment (JakeWharton) [inline redwood-lazylayout-view/src/main/kotlin/app/cash/redwood/lazylayout/view/ViewLazyList.kt:137]: Looks like this went from truncating to rounding and changed the snapshot tests in a non-trivial way. Needs recaptured.

Comment (swankjesse) [reply redwood-lazylayout-view/src/main/kotlin/app/cash/redwood/lazylayout/view/ViewLazyList.kt:137]: Yep, that’s it. Good catch

**Review by underscoretang — APPROVED** [review]

**Pull request merged**